File: convert.py
import os
import struct
import zlib
import base64

# Thư mục đầu vào và đầu ra
INPUT_FOLDER = "hls"
OUTPUT_FOLDER = "hls_png"
M3U8_FILE = os.path.join(INPUT_FOLDER, "playlist.m3u8")  # File m3u8 gốc
OUTPUT_M3U8_FILE = os.path.join(OUTPUT_FOLDER, "playlist.m3u8")  # File m3u8 mới

# Tạo thư mục đầu ra nếu chưa tồn tại
os.makedirs(OUTPUT_FOLDER, exist_ok=True)

def create_png_from_ts(ts_path, png_path):
    """Nhúng file TS vào file PNG"""
    png_header = b"\x89PNG\r\n\x1a\n"
    ihdr_data = b"\x00\x00\x00\x01\x00\x00\x00\x01\x08\x06\x00\x00\x00"
    ihdr_crc = struct.pack(">I", zlib.crc32(b"IHDR" + ihdr_data))
    ihdr_chunk = b"\x00\x00\x00\x0D" + b"IHDR" + ihdr_data + ihdr_crc
    idat_data = b"\x78\x9C\x63\x00\x01\x00\x00\x05\x00\x01"
    idat_crc = struct.pack(">I", zlib.crc32(b"IDAT" + idat_data))
    idat_chunk = struct.pack(">I", len(idat_data)) + b"IDAT" + idat_data + idat_crc
    iend_chunk = b"\x00\x00\x00\x00IEND\xAE\x42\x60\x82"
    
    with open(ts_path, "rb") as f:
        ts_data = f.read()
    
    with open(png_path, "wb") as f:
        f.write(png_header + ihdr_chunk + idat_chunk + iend_chunk + ts_data)

def update_m3u8():
    """Cập nhật file m3u8, thay thế .ts bằng .png"""
    if not os.path.exists(M3U8_FILE):
        print(f"⚠️ Không tìm thấy {M3U8_FILE}, bỏ qua bước cập nhật m3u8.")
        return
    
    with open(M3U8_FILE, "r", encoding="utf-8") as f:
        lines = f.readlines()
    
    new_lines = []
    for line in lines:
        line = line.strip()
        if line.endswith(".ts"):
            png_name = os.path.splitext(os.path.basename(line))[0] + ".png"
            new_lines.append(png_name)
        else:
            new_lines.append(line)
    
    with open(OUTPUT_M3U8_FILE, "w", encoding="utf-8") as f:
        f.write("\n".join(new_lines) + "\n")
    print(f"✅ Đã cập nhật {OUTPUT_M3U8_FILE}")

# Duyệt qua folder hls
for file in os.listdir(INPUT_FOLDER):
    if file.endswith(".ts"):
        ts_path = os.path.join(INPUT_FOLDER, file)
        png_path = os.path.join(OUTPUT_FOLDER, os.path.splitext(file)[0] + ".png")
        create_png_from_ts(ts_path, png_path)
        print(f"✅ Đã chuyển {file} -> {os.path.basename(png_path)}")

# Cập nhật file m3u8
update_m3u8()
print("🎉 Hoàn tất chuyển đổi!")

File: trick/convert.js
const fs = require("fs");
const path = require("path");
const { PNG } = require("pngjs");

const INPUT_FOLDER = "hls";
const OUTPUT_FOLDER = "hls_png";
const M3U8_FILE = path.join(INPUT_FOLDER, "playlist.m3u8");
const OUTPUT_M3U8_FILE = path.join(OUTPUT_FOLDER, "playlist.m3u8");

// Tạo thư mục output nếu chưa có
if (!fs.existsSync(OUTPUT_FOLDER)) {
  fs.mkdirSync(OUTPUT_FOLDER, { recursive: true });
}

function createPngFromTs(tsPath, pngPath) {
  const tsData = fs.readFileSync(tsPath); // Đọc file .ts

  // Tạo ảnh PNG 1x1 pixel
  const png = new PNG({
    width: 1,
    height: 1,
    filterType: 4,
  });

  png.data[0] = 255; // Red
  png.data[1] = 255; // Green
  png.data[2] = 255; // Blue
  png.data[3] = 255; // Alpha

  // Ghi file PNG và nhúng dữ liệu TS vào
  const writeStream = fs.createWriteStream(pngPath);
  png.pack().pipe(writeStream);

  writeStream.on("finish", () => {
    fs.appendFileSync(pngPath, tsData); // Nhúng dữ liệu TS vào PNG
    console.log(`✅ Đã chuyển ${path.basename(tsPath)} -> ${path.basename(pngPath)}`);
  });
}

function updateM3U8() {
  if (!fs.existsSync(M3U8_FILE)) {
    console.warn(`⚠️ Không tìm thấy ${M3U8_FILE}, bỏ qua bước cập nhật m3u8.`);
    return;
  }

  const lines = fs.readFileSync(M3U8_FILE, "utf8").split("\n");
  const newLines = lines.map(line => {
    if (line.endsWith(".ts")) {
      return line.replace(".ts", ".png");
    }
    return line;
  });

  fs.writeFileSync(OUTPUT_M3U8_FILE, newLines.join("\n"));
  console.log(`✅ Đã cập nhật ${OUTPUT_M3U8_FILE}`);
}

// Duyệt qua thư mục input và chuyển đổi file
fs.readdirSync(INPUT_FOLDER).forEach(file => {
  if (file.endsWith(".ts")) {
    const tsPath = path.join(INPUT_FOLDER, file);
    const pngPath = path.join(OUTPUT_FOLDER, file.replace(".ts", ".png"));
    createPngFromTs(tsPath, pngPath);
  }
});

// Cập nhật file M3U8
updateM3U8();
console.log("🎉 Hoàn tất chuyển đổi!");
